legal-api/src/legal_api/services/authz.py

```diff
@@ -95,96 +107,441 @@ def has_roles(jwt: JwtManager, roles: List[str]) -> bool:
+                if has_blocker(business, state_filing, allowable_filing_value, has_blocker_default):
+                    continue
+                allowable_filing_types \
+                    .append({'name': allowable_filing_key,
+                             'displayName': FilingMeta.get_display_name(legal_type, allowable_filing_key),
+                             'feeCode': Filing.get_fee_code(legal_type, allowable_filing_key)})
+        else:
+            filing_sub_type_items = \
+                filter(lambda x: legal_type in x[1].get('legalTypes', []), allowable_filing_value.items())
+            for filing_sub_type_item_key, filing_sub_type_item_value in filing_sub_type_items:
+                if has_blocker(business, state_filing, filing_sub_type_item_value, has_blocker_default):
+                    continue
+                allowable_filing_types \
+                    .append({'name': allowable_filing_key,
+                             'type': filing_sub_type_item_key,
+                             'displayName': FilingMeta.get_display_name(legal_type,
+                                                                        allowable_filing_key,
+                                                                        filing_sub_type_item_key),
+                             'feeCode': Filing.get_fee_code(legal_type,
+                                                            allowable_filing_key,
+                                                            filing_sub_type_item_key)})
+
+    return allowable_filing_types
+
+
+def has_blocker(business: Business, state_filing: Filing, allowable_filing: dict, has_default_blocker: bool):
+    """Return True if allowable filing has a blocker."""
+    if not (blocker_checks := allowable_filing.get('blockerChecks', {})):
+        return False
+
+    if has_business_blocker_default(blocker_checks, has_default_blocker):
+        return True
+
+    if has_blocker_valid_state_filing(state_filing, blocker_checks):
+        return True
+
+    if has_blocker_invalid_state_filing(state_filing, blocker_checks):
+        return True
+
+    if has_blocker_warning_filing(business.warnings, blocker_checks):
+        return True
+
+    return False
+
+
+def has_business_blocker_default(blocker_checks: dict, has_default_blocker: bool):
+    """Return True if the business has a default blocker."""
+    if not (business_blocker_checks := blocker_checks.get('business', [])):
+        return False
+
+    if BusinessBlocker.DEFAULT not in business_blocker_checks:
+        return False
+
+    return has_default_blocker
+
+
+def business_blocker_default_check(business: Business):
+    """Return True if the business has a default blocker condition."""
+    if not business:
+        return False
+
+    if has_blocker_filing(business):
+        return True
+
+    if business.admin_freeze:
+        return True
+
+    return False
+
+
+def has_blocker_filing(business: Business):
```

Comment:
Have you thought about blocker tasks, eg, "Annual Report is due"... what does that look like?

Reply:
From what I could see, the FE was only really looking for pending filings as blockers at the end of the day. 

So when an AR to do is on the entity dashboard, it isn't blocking any filings.  But once an AR filing is initiated, my logic is it will move into DRAFT which should be picked up by the `BusinessBlocker.DEFAULT` check. 

Let me know if you see any problems with this approach.

Reply:
Yes, you're right -- an Annual Report task (ie, due) does not block any other filings.

What's the other task we could have -- conversion? Do you have a business I could look at?

Reply:
Yes, a conversion task (todo) blocks other filings. Does the BE handle this?

Eg, https://dev.bcregistry.ca/business/FM1000007

Reply:
I don't specifically check for a conversion to do.  But `WarningType.MISSING_REQUIRED_BUSINESS_INFO` is essentially the conversion check.

Reply:
^^ yes, if there is a compliance warning or missing info warning then that's a blocker. Do you handle that?

Reply:
yes, if you take a look at this test, no filing actions are returned to the the user.

https://github.com/bcgov/lear/blob/7616ed51b1430c633187e55f7986a02eea904496/legal-api/tests/unit/services/test_authorization.py#L1354

For staff user, it does return some filing actions:
https://github.com/bcgov/lear/blob/7616ed51b1430c633187e55f7986a02eea904496/legal-api/tests/unit/services/test_authorization.py#L1331-L1336

Reply:
Awesome 👍